server/src/main/java/org/apache/druid/segment/realtime/firehose/CombiningFirehoseFactory.java

```diff
@@ -33,11 +36,12 @@
 import java.io.IOException;
 import java.util.Iterator;
 import java.util.List;
+import java.util.stream.Stream;
 
 /**
  * Creates firehose that combines data from different Firehoses. Useful for ingesting data from multiple sources.
  */
-public class CombiningFirehoseFactory implements FirehoseFactory<InputRowParser>
+public class CombiningFirehoseFactory implements FiniteFirehoseFactory<InputRowParser, List<FirehoseFactory>>
```

Comment:
I looked for implementations of `FirehoseFactory` that do not also implement `FiniteFirehoseFactory` and I found a few... eg. `ClippedFirehoseFactory`, `FixedCountFirehoseFactory`, etc.

I think the underlying problem is in `IndexIOConfig#getNonNullInputSource`

`IndexIOConfig` accepts a FirehoseFactory, but [getNonNullInputSource](https://github.com/apache/druid/blob/f6594fff608d4b2e071c7bdd6d86d7f87398ce4f/indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java#L1135) expects a `FiniteFirehoseFactory`.

I wonder if we could fix this by creating an adapter that translates any FirehoseFactory into a non-splittable FiniteFirehoseFactory (ie. the same functionality you've implemented in this class).

I understand that the FirehoseFactories are deprecated, so I'm open to other ways of fixing the broken quickstart material as well - can we update the spec used so that the combining firehose factory works without a code change?

Reply:
I feel that adding a non-splittable adapter would apply only to `CombiningFirehoseFactory`.  `ClippedFirehoseFactory` , `TimedShutoffFirehoseFactory` and `FixedCountFirehoseFactory` can be potentially splittable if the delegate `FirehoseFactory` is splittable, it's just that they haven't been fixed yet. 
What we actually need is for index_parallel to support nested InputSource say DelegateInputSource. The delegate FirehoseFactories such as the ones listed above can then be easily provided their equivalent InputSource implementations. I've been looking at adding this support and hopefully can have a proposal out once I have something.